File: ObjC/Restaurant/Classes/ReservationViewController.m
//
//  ReservationViewController.m
//  Restaurants
//
//  Created by AppsFoundation on 1/25/15.
//  Copyright (c) 2015 AppsFoundation. All rights reserved.
//

#import "ReservationViewController.h"

//cells
#import "ReservationLocationTableViewCell.h"
#import "NumberOfGuestsTableViewCell.h"
#import "PhoneNumberTableViewCell.h"
#import "MakeReservationTableViewCell.h"

typedef enum {
    ReservationCellLocation = 0,
    ReservationCellNumberOfGuests,
    ReservationCellPhone,
    ReservationCellMakeReservation,
    ReservationCellCount
}ReservationCell;

@interface ReservationViewController ()

@end

@implementation ReservationViewController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return ReservationCellCount;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    switch (indexPath.row) {
        case ReservationCellLocation: {
            ReservationLocationTableViewCell *cell = (ReservationLocationTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"reservationCell"];
            cell.backgroundColor = [UIColor clearColor];
            return cell;
        }
        case ReservationCellNumberOfGuests: {
            NumberOfGuestsTableViewCell *cell = (NumberOfGuestsTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"numberOfGuestsCell"];
            cell.backgroundColor = [UIColor clearColor];
            return cell;
        }
        case ReservationCellPhone: {
            PhoneNumberTableViewCell *cell = (PhoneNumberTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"phoneNumberCell"];
            cell.backgroundColor = [UIColor clearColor];
            return cell;
        }
        case ReservationCellMakeReservation: {
            MakeReservationTableViewCell *cell = (MakeReservationTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"makeReservationCell"];
            cell.backgroundColor = [UIColor clearColor];
            return cell;
        }
    }
    
    return nil;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == ReservationCellMakeReservation) {
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Thank You" message:@"You have booked table. Thanks for your reservation." delegate:nil cancelButtonTitle:@"Ok" otherButtonTitles:nil];
        [alert show];
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    switch (indexPath.row) {
        case ReservationCellLocation: {
            return 228.0f;
        }
        case ReservationCellNumberOfGuests: {
            return 140.0f;
        }
        case ReservationCellPhone: {
            return 140.0f;
        }
        case ReservationCellMakeReservation: {
            return 98.0f;
        }
    }
    
    return 0.0f;
}

@end

File: Swift/Restaurant/Classes/ReservationViewController.swift
//
//  ReservationViewController.swift
//  Restaurant
//
//  Created by AppsFoundation on 8/28/15.
//  Copyright © 2015 AppsFoundation. All rights reserved.
//

import UIKit

enum ReservationCell: Int {
    case Location = 0
    case NumberOfGuests
    case Phone
    case MakeReservation
    case Count
}

class ReservationViewController: BaseViewController {
    
    @IBOutlet weak var tableView: UITableView?
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
}

// MARK: - UITableViewDataSource
extension ReservationViewController: UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return ReservationCell.Count.rawValue
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch indexPath.row {
        case ReservationCell.Location.rawValue:
            let cell: ReservationLocationTableViewCell = tableView.dequeueReusableCell(withIdentifier: "reservationCell") as! ReservationLocationTableViewCell
            cell.backgroundColor = UIColor.clear
            return cell
        case ReservationCell.NumberOfGuests.rawValue:
            let cell: NumberOfGuestsTableViewCell = tableView.dequeueReusableCell(withIdentifier: "numberOfGuestsCell") as! NumberOfGuestsTableViewCell
            cell.backgroundColor = UIColor.clear
            return cell
        case ReservationCell.Phone.rawValue:
            let cell: PhoneNumberTableViewCell = tableView.dequeueReusableCell(withIdentifier: "phoneNumberCell") as! PhoneNumberTableViewCell
            cell.backgroundColor = UIColor.clear
            return cell
        case ReservationCell.MakeReservation.rawValue:
            let cell: MakeReservationTableViewCell = tableView.dequeueReusableCell(withIdentifier: "makeReservationCell") as! MakeReservationTableViewCell
            cell.backgroundColor = UIColor.clear
            return cell
        default:
            return UITableViewCell() //return empty cell
        }
    }
}

// MARK: - UITableViewDelegate
extension ReservationViewController: UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if indexPath.row == ReservationCell.MakeReservation.rawValue {
            let alert = UIAlertView(title: "Thank You", message: "You have booked table. Thanks for your reservation.", delegate: nil, cancelButtonTitle: "Ok")
            alert.show()
        }
    }
}
